### PR TITLE
k8s/watchers: fix data race in (*K8sWatcher).addK8sServiceV1

### LINKED ISSUE #13603: [RACE detector] data race in pkg/k8s/watchers.(*K8sWatcher).addK8sServiceV1
```
2020-10-16T13:03:52.869354463Z WARNING: DATA RACE
2020-10-16T13:03:52.869356850Z Read at 0x00c00000c350 by goroutine 104:
2020-10-16T13:03:52.869358989Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addK8sServiceV1()
2020-10-16T13:03:52.869361156Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/service.go:87 +0x17b
2020-10-16T13:03:52.869363212Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).servicesInit.func1()
2020-10-16T13:03:52.869365226Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/service.go:42 +0x124
2020-10-16T13:03:52.869367176Z   k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd()
2020-10-16T13:03:52.869369173Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x8f
2020-10-16T13:03:52.869371199Z   k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
2020-10-16T13:03:52.869373170Z       <autogenerated>:1 +0x2a
2020-10-16T13:03:52.869375315Z   github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1()
2020-10-16T13:03:52.869377316Z       /go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:119 +0x2a9
2020-10-16T13:03:52.869379300Z   k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop()
2020-10-16T13:03:52.869381320Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:501 +0x3ab
2020-10-16T13:03:52.869383381Z   k8s.io/client-go/tools/cache.(*controller).processLoop()
2020-10-16T13:03:52.869385351Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:179 +0x83
2020-10-16T13:03:52.869387356Z   k8s.io/client-go/tools/cache.(*controller).processLoop-fm()
2020-10-16T13:03:52.869389313Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:177 +0x44
2020-10-16T13:03:52.869391332Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
2020-10-16T13:03:52.869393347Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
2020-10-16T13:03:52.869403569Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
2020-10-16T13:03:52.869405719Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xba
2020-10-16T13:03:52.869407742Z   k8s.io/apimachinery/pkg/util/wait.JitterUntil()
2020-10-16T13:03:52.869409692Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x114
2020-10-16T13:03:52.869411710Z   k8s.io/apimachinery/pkg/util/wait.Until()
2020-10-16T13:03:52.869413622Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4a7
2020-10-16T13:03:52.869415640Z   k8s.io/client-go/tools/cache.(*controller).Run()
2020-10-16T13:03:52.869417572Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:150 +0x449
2020-10-16T13:03:52.869419598Z 
2020-10-16T13:03:52.869421920Z Previous write at 0x00c00000c350 by goroutine 144:
2020-10-16T13:03:52.869423937Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit.func1()
2020-10-16T13:03:52.869425915Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:118 +0x22b
2020-10-16T13:03:52.869427902Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit()
2020-10-16T13:03:52.869429857Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:138 +0x7ef
2020-10-16T13:03:52.869431831Z 
2020-10-16T13:03:52.869434204Z Goroutine 104 (running) created at:
2020-10-16T13:03:52.869434204Z Goroutine 104 (running) created at:
2020-10-16T13:03:52.869354463Z WARNING: DATA RACE
2020-10-16T13:03:52.869356850Z Read at 0x00c00000c350 by goroutine 104:
2020-10-16T13:03:52.869358989Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).addK8sServiceV1()
2020-10-16T13:03:52.869361156Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/service.go:87 +0x17b
2020-10-16T13:03:52.869363212Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).servicesInit.func1()
2020-10-16T13:03:52.869365226Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/service.go:42 +0x124
2020-10-16T13:03:52.869367176Z   k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd()
2020-10-16T13:03:52.869369173Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:227 +0x8f
2020-10-16T13:03:52.869371199Z   k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
2020-10-16T13:03:52.869373170Z       <autogenerated>:1 +0x2a
2020-10-16T13:03:52.869375315Z   github.com/cilium/cilium/pkg/k8s/informer.NewInformerWithStore.func1()
2020-10-16T13:03:52.869377316Z       /go/src/github.com/cilium/cilium/pkg/k8s/informer/informer.go:119 +0x2a9
2020-10-16T13:03:52.869379300Z   k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop()
2020-10-16T13:03:52.869381320Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/delta_fifo.go:501 +0x3ab
2020-10-16T13:03:52.869383381Z   k8s.io/client-go/tools/cache.(*controller).processLoop()
2020-10-16T13:03:52.869385351Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:179 +0x83
2020-10-16T13:03:52.869387356Z   k8s.io/client-go/tools/cache.(*controller).processLoop-fm()
2020-10-16T13:03:52.869389313Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:177 +0x44
2020-10-16T13:03:52.869391332Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
2020-10-16T13:03:52.869393347Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x75
2020-10-16T13:03:52.869403569Z   k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
2020-10-16T13:03:52.869405719Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xba
2020-10-16T13:03:52.869407742Z   k8s.io/apimachinery/pkg/util/wait.JitterUntil()
2020-10-16T13:03:52.869409692Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x114
2020-10-16T13:03:52.869411710Z   k8s.io/apimachinery/pkg/util/wait.Until()
2020-10-16T13:03:52.869413622Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x4a7
2020-10-16T13:03:52.869415640Z   k8s.io/client-go/tools/cache.(*controller).Run()
2020-10-16T13:03:52.869417572Z       /go/src/github.com/cilium/cilium/vendor/k8s.io/client-go/tools/cache/controller.go:150 +0x449
2020-10-16T13:03:52.869419598Z 
2020-10-16T13:03:52.869421920Z Previous write at 0x00c00000c350 by goroutine 144:
2020-10-16T13:03:52.869423937Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit.func1()
2020-10-16T13:03:52.869425915Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:118 +0x22b
2020-10-16T13:03:52.869427902Z   github.com/cilium/cilium/pkg/k8s/watchers.(*K8sWatcher).podsInit()
2020-10-16T13:03:52.869429857Z       /go/src/github.com/cilium/cilium/pkg/k8s/watchers/pod.go:138 +0x7ef
2020-10-16T13:03:52.869431831Z 
2020-10-16T13:03:52.869434204Z Goroutine 104 (running) created at:

```

https://jenkins.cilium.io/view/Race%20Detection/job/Ginkgo-CI-Tests-k8s1.19-Pipeline-Race-Detection/11/

### PR DESCRIPTION
**Pull request by tklauser**
Access to k.podStore must be protected by k.podStoreMU.

Closes #13603

Fixes: e7bb8a7eadb5 ("k8s/cilium Event handlers and processing logic for LRPs")

### CODE REVIEW

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
test-race-4.9

**Review by aanm — CHANGES_REQUESTED** [review]
@tklauser I'm fine with the changes but this should not be the real fix. This code architecture should be done differently as well.

@aditighag `k.podStore` should not be passed for every k8s service add, we should register a store in https://github.com/cilium/cilium/blob/e91f803c360484dc109eb0d18badc1afa92d457f/daemon/cmd/daemon.go#L367 and use it as follows:

```patch
diff --git a/daemon/cmd/daemon.go b/daemon/cmd/daemon.go
index 8c334acaa..faae4336b 100644
--- a/daemon/cmd/daemon.go
+++ b/daemon/cmd/daemon.go
@@ -365,6 +365,7 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
        )
 
        d.redirectPolicyManager.RegisterSvcCache(&d.k8sWatcher.K8sSvcCache)
+       d.redirectPolicyManager.RegisterGetStores(&d.k8sWatcher)
 
        bootstrapStats.daemonInit.End(true)
 
diff --git a/pkg/redirectpolicy/manager.go b/pkg/redirectpolicy/manager.go
index 5213fcb30..bfce4b013 100644
--- a/pkg/redirectpolicy/manager.go
+++ b/pkg/redirectpolicy/manager.go
@@ -65,6 +65,8 @@ type Manager struct {
 
        svcCache svcCache
 
+       storeGetter StoreGetter
+
        // Mutex to protect against concurrent access to the maps
        mutex lock.Mutex
 
@@ -95,6 +97,14 @@ func (rpm *Manager) RegisterSvcCache(cache svcCache) {
        rpm.svcCache = cache
 }
 
+type StoreGetter interface {
+       GetStore(name string) cache.Store
+}
+
+func (rpm *Manager) RegisterGetStores(sg StoreGetter) {
+       rpm.storeGetter = sg
+}
+
 // Event handlers
 
 // AddRedirectPolicy parses the given local redirect policy config, and updates
@@ -503,12 +513,13 @@ func (rpm *Manager) upsertService(config *LRPConfig, frontendMapping *feMapping)
 }
 
 // Returns a slice of endpoint pods metadata that are selected by the given policy config.
-func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig, podStore cache.Store) []*podMetadata {
+func (rpm *Manager) getLocalPodsForPolicy(config *LRPConfig) []*podMetadata {
        var (
                retPods []*podMetadata
                podData *podMetadata
        )
 
+       podStore := rpm.storeGetter.GetStore("pod")
        for _, podItem := range podStore.List() {
                pod, ok := podItem.(*slimcorev1.Pod)
                if !ok || !config.checkNamespace(pod.GetNamespace()) {

```

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
Updated as per @aanm's suggestion (https://github.com/cilium/cilium/pull/13604#pullrequestreview-510554177), see commit ea208e7d1796c38163de1ae35c488b1f08967878

File: pkg/redirectpolicy/manager_test.go [hunk 1]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/cilium/cilium/pkg/k8s/utils"
 	lb "github.com/cilium/cilium/pkg/loadbalancer"
 	"github.com/cilium/cilium/pkg/policy/api"
+	"k8s.io/client-go/tools/cache"
```
Comment (christarazi) [inline pkg/redirectpolicy/manager_test.go]: Nit: import ordering

**Review by christarazi — APPROVED** [review]
Minor nit

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aditighag** [comment]
Thanks for fixing this, @tklauser and @aanm.

**Pull request merged**